File: Eyes/ImageTester/src/lib/java/com/applitools/ImageTester/TestObjects/TestUnit.java
package com.applitools.ImageTester.TestObjects;

import com.applitools.ImageTester.Interfaces.IDisposable;
import com.applitools.ImageTester.Interfaces.ITestable;

import java.io.File;

public abstract class TestUnit implements ITestable, IDisposable {
    protected final File file_;
    protected String name_;

    protected TestUnit(File file) {
        file_ = file;
        if (file != null && !file.canRead()) {
            throw new RuntimeException(String.format("Unreadable path/file %s, might be a permission issue!", file.getAbsolutePath()));
        }
    }

    protected TestUnit(String name) {
        file_ = null;
        name_ = name;
    }

    public String name() {
        return file_ == null ? name_ : file_.getName();
    }

    public File getFile() {
        return file_;
    }

}

File: Eyes/ImageTester/src/csharp/lib/TestObjects/TestUnit.cs
using System;
using System.IO;
using Applitools.Images;
using Applitools.ImageTester.Interfaces;

namespace Applitools.ImageTester.TestObjects
{
    public abstract class TestUnit : ITestable
    {
        protected readonly FileStream file;
        protected string name;

        protected TestUnit(FileStream file)
        {
            this.file = file;
            if (file != null && !file.CanRead)
            {
                throw new InvalidOperationException(string.Format("Unreadable path/file '{0}', might be a permission issue!", file.Name));
            }
        }

        protected TestUnit(string name)
        {
            file = null;
            this.name = name;
        }

        public virtual string Name()
        {
            return file == null ? name : file.Name;
        }

        public virtual FileStream GetFile()
        {
            return file;
        }

        public abstract void Run(Eyes eyes);

        public abstract void Dispose();
    }
}